plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/AutoTaggingActionFilter.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.wlm;
+
+import org.opensearch.action.ActionRequest;
+import org.opensearch.action.IndicesRequest;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchScrollRequest;
+import org.opensearch.action.support.ActionFilter;
+import org.opensearch.action.support.ActionFilterChain;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.action.ActionResponse;
+import org.opensearch.plugin.wlm.rule.attribute_extractor.IndicesExtractor;
+import org.opensearch.rule.InMemoryRuleProcessingService;
+import org.opensearch.tasks.Task;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.wlm.WorkloadGroupTask;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * This class is responsible to evaluate and assign the QUERY_GROUP_ID header in ThreadContext
+ */
+public class AutoTaggingActionFilter implements ActionFilter {
```

Comment:
Will every plugin write their own action filter? As the logic looks pretty common to me.

Reply:
The ActionFilter logic is use case specific hence can't be generalised. Although it looks common but the use cases could be different, for this case we are populating ThreadContext with WLM specific header but for other it could very well be just a point in the execution path to do some metrices/logging related stuff